Library/MeasurePlugin.h

```diff
@@ -39,6 +40,8 @@ class MeasurePlugin : public Measure
 	virtual UINT GetTypeID() { return TypeID<MeasurePlugin>(); }
 
 	virtual const WCHAR* GetStringValue();
+	const WCHAR* GetImageData(PluginImageData &imageData);
+	bool HasImageData() { return m_GetImageFunc != nullptr; }

```

Comment:
The fact that plugin has function with name `GetImage` doesn't mean that it can provide images.
1. It can be an old plugin, that uses `GetImage` just as a section variable.
2. If it's a new plugin, not all of the instances may provide images. Like, for example, in AudioAnalyzer parent measure will never provide images.
3. Even if an instance provides image value, it may stop providing it in any moment, when measure configuration changes, or when something inside the plugin changes.
For example, in AudioAnalyzer someone can change parent measure configuration, and remove the thing that was generating images, and child measures will stop providing the images. Or just audio device configuration changes, and something becomes invalid, and images can no longer be generated.
Or you could imagine a million other cases for when something inside the plugin prevents a measure from providing an image.

Reply:
I think this function should be removed, and we should check whether the plugin filled the image on each call to `GetImageData`.

Reply:
But this is not a section variable that a user can call, this is similar to `GetString`, and i doubt that another plugin uses it with this signature.

>Considering the fact that GUI should ideally not know anything about backend (even if it's already not true for rainmeter), I think it would be better here to declare function like this:

Btw, this gave me an idea, i don't think we need `PluginImageData` at all in this case, i can just make more arguments in all function that need this struct, and users can create there own custom stucts, they just need to provide data with these types `uint8_t`, `int`, `int`, `uint64_t`.

Reply:
> But this is not a section variable that a user can call

It's not a section variable _now_, _after your changes_.
It's a perfectly valid section variable in the official rainmeter branch, so plugins can use it, and we really want to make sure that we don't break anything, if we want in-memory transfer to be accepted.

> i doubt that another plugin uses it with this signature

The issue is: it doesn't matter it anybody used this, because it's a part of official rainmeter plugin API, and it was _valid_ to use this.

I hope you understand now why I _hate_ Rainmeter implementation of user-defined-functions aka section variables so-so-so much. It's soooo hard to add _anything_ to the plugin api.
And the worst thing is that it's actually is inconvenient (remember asking me why `resolve` doesn't work with any case? that's why) and is absolutely useless, because rainmeter would lose nothing if it didn't say "100% of functions that we don't define can be used as a section variables" and just defined something like `ResolveSectionVariable` function into plugin api.

> i can just make more arguments in all function that need this struct, and users can create there own custom stucts

Well, yeah. If it doesn't look terrible because of too many arguments.

Reply:
>I think this function should be removed, and we should check whether the plugin filled the image on each call to `GetImageData`.

But howw can it fill the image and that function is the only way to access it?

>Well, yeah. If it doesn't look terrible because of too many arguments.

```cpp
PLUGIN_EXPORT LPCWSTR GetImage(void *data, UINT8 pixels, INT32 width, INT32 height, UINT64 timestamp);
```

My second main reason for is was what you mentioned above.
But except removing that function, isn't there a way to make this function unique? i mean changing the name or signature or any trick.

Reply:
>But howw can it fill the image and that function is the only way to access it?

I don't quite understand what you mean. Could you rephrase it?

>`PLUGIN_EXPORT LPCWSTR GetImage(void *data, UINT8 pixels, INT32 width, INT32 height, UINT64 timestamp);`

Oh, changing signature of this function is a different thing.
Whether we _can_ change it depends on what we do with the APIs.

But let's talk about would it be better or not.
To be honest, I don't give a shit about how api functions _look_, because I almost never use them. In my plugins I only use C++ wrappers that are so much more convenient than raw C API that we have.
I see 2 considerations regarding this:
1. Struct layout. We have to make sure that the struct layout is stable. Same thing as types: when we change compiler or some options of the compiler, we have to make sure that struct layout doesn't change.
And it's not obvious what we need to do this. It would require additional research in the google.
2. Extensibility. Imagine we didn't have the timestamp, and were recalculating everything on each update, and then we thought: oh, it would be cool optimize it. But how? If we were to compare all the pixels on each update, it probably wouldn't be a good optimization. So ideally we would want the plugin to tell us if the image has changed. So we would need to add a new argument.
We can't really add arguments to functions of a public api, because it would make it incompatible. Well, maybe we could do something cheesy, but it's really not what you want to be doing with public APIs.
However, in case of structs, we can just add a new field, make sure that layout of old fields didn't change and the plugin knows it can use the new field, and everybody is happy.
Would we ever need extensibility in case of image transfer? Who knows. But if we need it but can't extend the api, it will be very sad.

I guess we can also add an optional argument, like `void*` type, and use it in the future if we ever need to extend the api.

Reply:
>I don't quite understand what you mean. Could you rephrase it?

If we removed `GetImage` function, how can the image meter request the image data to be filled?

>Struct layout. We have to make sure that the struct layout is stable.

What does struct layout mean? :P

>Would we ever need extensibility in case of image transfer? Who knows. But if we need it but can't extend the api, it will be very sad.

omg, then i'll stick with my lovely `PluginImageData` struct.
Wait, omg2 because this also can't be extended ☹

>However, in case of structs, we can just add a new field, make sure that layout of old fields didn't change and the plugin knows it can use the new field, and everybody is happy.

hmm, i didn't get it, wdum by field?

>I guess we can also add an optional argument, like void* type, and use it in the future if we ever need to extend the api.

If i get it correctly, any api changes require changes to rainmeter as well, because how are we going to use `void*` if rainmeter doesn't know what to do with it?

Reply:
> If we removed `GetImage` function, how can the image meter request the image data to be filled?

I meant that only `HasImageData` should be removed. `GetImage` is fine.

> What does struct layout mean? :P

It means how the struct fields are actually placed in the memory.
Like, their order (easy, it usually just corresponds to their order in struct definition), and not much empty space is between them (hard, depends on compiler settings).

> Wait, omg2 because this also can't be extended ☹

It could be extended if you added something like `int8_t structVersion` field, for example.
Either way, after some thoughts I think that I like the solution with `void*` more.

> hmm, i didn't get it, wdum by field?

```cpp
struct MyStruct {
int x,
float y;
};
```
`x` and `y` are fields of this struct.

> If i get it correctly, any api changes require changes to rainmeter as well, because how are we going to use `void*` if rainmeter doesn't know what to do with it?

Yes. Any api _semantics_ changes changes would require changes in rainmeter.
Good thing about this is that when semantics changes, the physical api doesn't change, so we break neither "old rainmeter + new plugin" case nor "new rainmeter + old plugin" case.
Like, the old function definition had `void Func(int x, void* reserved)`, and new has `void Func(int x, void* extension)`. Function signature is the same, so we don't have to brainstorm how to change it without breaking some unknown plugins.

Reply:
>Like, their order (easy, it usually just corresponds to their order in struct definition), and not much empty space is between them (hard, depends on compiler settings).

Something like padding? if yes then i'm in trouble, it's one of the things that i still dunno how to deal with.

>Either way, after some thoughts I think that I like the solution with void* more.

>Like, the old function definition had void Func(int x, void* reserved), and new has void Func(int x, void* extension). Function signature is the same, so we don't have to brainstorm how to change it without breaking some unknown plugins.

```cpp
PLUGIN_EXPORT LPCWSTR GetImage(void *data, PluginImageData imageData, void* reserved);
```

Should we go with this?

Reply:
Yeah, something like padding.
I too don't have an experience dealing with it.

> ```c++
> PLUGIN_EXPORT LPCWSTR GetImage(void *data, PluginImageData imageData, void* reserved);
> ```

We would probably want to use separate arguments:
```c++
PLUGIN_EXPORT int32_t GetImage(void *data, *pixels, *width, *height, void* reserved);
```
Also _if_ we decide to break the compatibility with section variables/custom functions, then we should also use a sane return value. Like, `int32_t` or `DWORD`.

Reply:
>Also if we decide to break the compatibility with section variables/custom functions, then we should also use a sane return value. Like, `int32_t` or `DWORD`.

"break the compatibility with section variables/custom functions"

Will this allow us to:

- Use old plugins in new rainmeter without function naming conflicts (the issue with `GetImage`)?
- Use new plugins in old rainmeter and still be able to tell the users to update there rainmeter?

That is pretty much all we want.

Reply:
https://stackoverflow.com/a/5398498

```PLUGIN_EXPORT int32_t GetImage(void *data, *pixels, *width, *height, void* reserved);```

i will go with this and avoid using structs, altho i compiled rainmeter using msvc but compiled the plugin with clang, i guess it made no difference since msvc is just a front end for llvm (just like clang afaik).

Reply:
> * Use new plugins in old rainmeter and still be able to tell the users to update there rainmeter?

No, new plugins just won't load in the old rainmeter, with log message like `can't load plugin (error 126)`.

Reply:
>No, new plugins just won't load in the old rainmeter, with log message like can't load plugin (error 126).

wait what? why it won't load the plugin :| i mean image meter should just normally read the string value of the measure.

>Also if we decide to break the compatibility with section variables/custom functions, then we should also use a sane return value. Like, `int32_t` or `DWORD`.

I used `bool` for testing, but idk how should other types be beneficial, i mean if the image got filled successfully return true else false, but with numbers... idk.

I mean yeah Booleans are just numbers but why not use `bool` in first place?

Also about this:

>>Use old plugins in new rainmeter without function naming conflicts (the issue with GetImage)?

Is the return type considered a part of the function signature?

Reply:
> wait what? why it won't load the plugin :| i mean image meter should just normally read the string value of the measure.

Because new plugins will require a new function in rainmeter api. When they are loaded in an old version, they can't call the new function, because the new function doesn't exist. Plugin loader doesn't know how to solve this, so the plugin just can't be loaded.
Plugins can't do anything about it. The only solution is to not add new functions into the api. I mean: exported functions. Solutions like `rmget(function)` are ok because new functions aren't _exported_.

> I mean yeah Booleans are just numbers but why not use `bool` in first place?

The api is in pure C, and pure C doesn't have bool. The fact that modern compilers allow this doesn't make it a good api.

>Is the return type considered a part of the function signature?

Technically no, return type is not part of the signature in C/C++.
Though, if you are talking about plugin api: in dll format there are just no signatures at all. They are not checked. You only have function names.

Reply:
>Plugins can't do anything about it. The only solution is to not add new functions into the api.

ah, you mean the rainmeter.lib, got it now.

>Though, if you are talking about plugin api: in dll format there are just no signatures at all. They are not checked. You only have function names.

That is really surprising, i didn't know it.

---

Okay, just to get the confusion out of my head, before i make the pull request, what do you suggest me to do to fix these problems:

1. Function naming issue (`GetImage` case).
2. Loading the plugins in old rainmeters.

Reply:
Nvm, we will discuss this later on in another issue.

Reply:
I suggest the following:
Extend RmGet to extend the api workout breaking  plugin loading.
Use new SetConfig to enable new plugin api. If a plugin called setconfig("rainmeter.plugin.images", "true") during initialization, then assume you can use getimage with new signature without fear, otherwise prevent loading images from this plugin and treat getimage as an old section variable.

Reply:
I don't wanna sound overwhelming.. but if possible, at a later time, could you open a new discussion for how to implement this? i got the general idea but still dunno about the code part, and really thank you so much in advanced. c^w^c

Reply:
Sure